cloudcredential/v1/types_aws.go

```diff
@@ -0,0 +1,62 @@
+/*
+Copyright 2018 The OpenShift Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// TODO: these types should eventually be broken out, along with the actuator, to a separate repo.
+
+// AWSProviderSpec contains the required information to create a user policy in AWS.
+// +k8s:deepcopy-gen=false
+type AWSProviderSpec struct {
+	metav1.TypeMeta `json:",inline"`
+	// StatementEntries contains a list of policy statements that should be associated with this credentials access key.
+	StatementEntries []StatementEntry `json:"statementEntries"`
+}
+
+// StatementEntry models an AWS policy statement entry.
+// +k8s:deepcopy-gen=false
+type StatementEntry struct {
+	// Effect indicates if this policy statement is to Allow or Deny.
+	Effect string `json:"effect"`
+	// Action describes the particular AWS service actions that should be allowed or denied. (i.e. ec2:StartInstances, iam:ChangePassword)
+	Action []string `json:"action"`
+	// Resource specifies the object(s) this statement should apply to. (or "*" for all)
+	Resource string `json:"resource"`
```

Comment:
should any of these be required?

Reply:
I suppose we could default to "*" if Resource is not set, but I'm not thrilled with the idea that by default we would be granting permissions with the least restrictions possible. If there were a workable minimum-permissions-by-default, I could definitely see making this not required.

Reply:
I don't think we need to require any of these, they seem optional, similarly how we do in audit. But defaulting to `*` is not a good idea. Default should be empty which means none.

Reply:
@soltysh sorry, getting back to this after some other unrelated work. If we mark this optional, and the empty string is treated as None, is there any value in defining a list of permissions and then saying that you can perform all those actions against nothing? What have we gained by making the field optional?

Reply:
Question: since this is an embedded type hidden behind a RawExtension, what effect does annotating these fields have?